[ADR] Define quality levels (#219)
apache/james-project: src/adr/0040-quality-levels-definitions.md

# 40. Quality levels definition

Date: 2020-06-21

## Status

Accepted (lazy consensus)

## Context

> James as a project delivers several artifacts, and features. In order for project users to better

Reviewer: What do you mean by artifacts here?

Author: `artifacts` => things James project users consumes.

 - Libraries as stated here: 
 - Mail Servers
 - Command line tools
 - Jars that can be used as an extension for a running James server

 - Decent integration tests coverage is needed
 - Performance tests need to be conducted out
 - Quality Assurance with external clients needs to be conducted out
 - known existing production deployments/usages
 - usable documentation

This is the maximum quality level delivered by the James project. Users should feel confident using these
artifacts or features.

By **experimental** we designate a part of the code not matching yet the above requirements. However some
active contributors are willing to raise the quality level of this component, and eventually make it 
supported. Or at least are willing to support users.

Users should have low expectations regarding experimental artifacts. They are encouraged to contribute to this 
component in order to raise its code quality.

By **unsupported** we mean that a component do not match most of the *supported* quality conditions. Active 
contributors do not feel confident delivering support for it. This artifact or feature might be deprecated and 
removed from future James releases. Users are strongly encouraged to contribute to the artifact development.

## Consequences

> Quality levels needs to be mentioned explicitly in the documentation, per product, and for each product, per feature.

Reviewer: ```suggestion
Quality levels need to be mentioned explicitly in the documentation, per profile, and for each profile, per feature.
```

Reviewer: I am not sure that a decision has been taken yet regarding the misleading term `product`, but I think propositions made like `profile` or `flavor` should maybe better used instead?

Author: Let's keep the word "artifact" instead of product/profile given that I correctly defines it.

# 40. Quality levels definition

Date: 2020-06-21

## Status

Accepted (lazy consensus)

## Context

We hereby define as an artifact compiled artifact that external people consumes. This includes:

 - libraries
 - Mail servers
 - Extensions for James Mail Servers
 - Command line tools

We designate as a feature an optional, opt-in behaviour of a James server that can be configured by 
user willing to rely on it.

James as a project delivers several artifacts, and features. In order for project users to better
understand the underlying quality of the artifact they use, as well as the level of risk associated,
we need to better define some quality levels.

## Decision

For a given artifact or feature, by **mature** we mean that:

>  - *interfaces* in components need a contract test suite

Reviewer 2: I would add: 'at least two implementations of the interface'

 - *implementation* of these interfaces need to pass this contract test suite which provides unit tests
 - Decent integration tests coverage is needed
 - Performance tests need to be conducted out
 - Quality Assurance with external clients needs to be conducted out
 - known existing production deployments/usages
 - usable documentation

This is the maximum quality level delivered by the James project. Users should feel confident using these
artifacts or features.

By **experimental** we designate a part of the code not matching yet the above requirements. However some
active contributors are willing to raise the quality level of this component, and eventually make it 
mature. Or at least are willing to support users.

Users should have low expectations regarding experimental artifacts or features. They are encouraged to contribute to them 

> in order to raise its code quality.

Reviewer 2: ```suggestion
in order to raise its quality.
```

# 40. Quality levels definition

Date: 2020-06-21

## Status

Accepted (lazy consensus)

## Context

We hereby define as an artifact compiled artifact that external people consumes. This includes:

 - libraries
 - Mail servers
 - Extensions for James Mail Servers
 - Command line tools

We designate as a feature an optional, opt-in behaviour of a James server that can be configured by 
user willing to rely on it.

> James as a project delivers several artifacts, and features. In order for project users to better

Reviewer 3: These seem to me like important and fundamental enough concepts to either expand the scope of this ADR, or to have a separate ADR that clearly defines (as a base concept for other ADRs like this one) the concepts of "artifact" and "feature".

Author: I think the context part does it well enough. If it is not clear enough, please provide alternative wordings.

Reviewer 3: That is not quite what I meant. I was not commenting on the words, I was commenting on the scope. This ADR is about "quality". However, the terms "artifacts" and "features" seem to be so fundamental that they should not be hidden away in a document that describes something different.

So we should either:

1. Expand the scope of this document (i.e. it defines "Artifacts", "Features", and "Quality Levels" and not just "Quality Levels", or
2. Have a separate ADR that defines "Artifacts" and "Features".

Then this can be referred to from other discussions which will surely use these fundamental concepts.

Just something to consider, maybe.

Author: I have a different approach here.

I started another document to explain which server we ship as part of the James project, then I hope that with @ieugen we come up with a list of libraries we deliver. Each one of them will get their own ADR, giving a de-facto definition (and exhaustive list of artifacts).

Regarding features, refining extension mechanisms and planning the work for their extraction would do the same.

understand the underlying quality of the artifact they use, as well as the level of risk associated,

> we need to better define some quality levels.

Reviewer 3: The document defines "quality levels" without defining "quality".

I think that "level" is obvious enough that it doesn't require any further definition, but "quality" may be interpreted differently by different people.

Author: > The document defines "quality levels" without defining "quality".

At one point it becomes hard for me to betterdefines such a generic term. I believes the levels definition disambiguate clearly this.

Reviewer 3: Writing is really hard, I agree. The hardest part is writing something "objective" that everybody understands the same way, especially in a project like this with people from different countries, backgrounds, cultures, education, generations... We'll never have a completely common understanding, so the trick is to decide the right point at which we should draw a boundary given the time and resources we have available. We will never completely agree on where that boundary should be. :)

Author: Actually this document, and the "mature" quality level dis-ambiguify what we mean by quality level, no?

Reviewer 3: Are you saying that "mature" implies "high quality"? In the definitions I proposed, a component could be mature, but unreliable.

I do agree that it's not easy figuring out how all the pieces fit together. Quality really depends on the objective. There is not "absolute". That is why context is so important.

## Decision

For a given artifact or feature, by **mature** we mean that:

>  - *interfaces* in components need a contract test suite

Reviewer 3: Would it be useful to describe "separation of API/implementation"? It is a term I have used a few times on the list, but I'm not sure I use it in the same way as others. (I have strongly corrupted by OSGi thinking, and it has a very clear meaning over there.)

"Interface" as you use it seems to suggest at the class level, whereas the real "interface" of a component is actualy at the package level, which may be why they call it "API" instead of "interface", to avoid confusion. Usually it is not just a single Java interface that defines behaviour, but rather an entire package. (Should generally not expand beyond a package.)

Author: > Would it be useful to describe "separation of API/implementation"?

:+1: and luckily it is the case almost everywhere.

> "Interface" as you use it seems to suggest at the class level, whereas the real "interface" of a component is actualy at the package level, which may be why they call it "API" instead of "interface", to avoid confusion. Usually it is not just a single Java interface that defines behaviour, but rather an entire package

If others don't complain, I will replace "interfaces" terms by "APIs".

 - *interfaces* have several implementations
 - *implementation* of these interfaces need to pass this contract test suite which provides unit tests
 - Decent integration tests coverage is needed
 - Performance tests need to be conducted out

>  - Quality Assurance with external clients needs to be conducted out

Reviewer 3: What do you mean by "Quality Assurance"? Is that defined in a different ADR?

Author: I mean "Manual Testing". Quality Assurance is in my opinion a common practice of the industry. Defining it is in my opinion not rely needed.

Reviewer 3: You may be surprised as to how something that **ought** to be clearly defined in the industry creates a lot of conflict. :)

If there is already a clear _de facto_ _modus operandi_ for the James community, or if nobody challenges it, then I guess it's fine. :)

 - known existing production deployments/usages
 - usable documentation

This is the maximum quality level delivered by the James project. Users should feel confident using these
artifacts or features.

By **experimental** we designate an artifact or feature not matching yet the above requirements. However some
active contributors are willing to raise the quality level of this component, and eventually make it 
mature. Or at least are willing to support users.

Users should have low expectations regarding experimental artifacts or features. They are encouraged to contribute to them 
in order to raise its quality.

> By **unsupported** we mean that an artifact or feature do not match most of the *mature* quality conditions. Active 

Reviewer 4: ```suggestion
By **unsupported** we mean that an artifact or feature does not match most of the *mature* quality conditions. Active 
```

> By **unsupported** we mean that an artifact or feature do not match most of the *mature* quality conditions. Active 

Reviewer 3: In the referenced email, @chibenwa wrote:

> This is a fair point, but 'quality' differs from 'support' concerns.

I agree with this point, which is why I have trouble with the "unsupported" label.

Author: Propose an alternative please. 

I want a term that reflexes the "nothing to expect" concepts. "unsupported" convey this idea quite well to be honest.

Reviewer 3: I wrote in a separate comment below how I think the concepts are a bit mixed up. But again, there comes a point at which a boundary has to be drawn, so even though IMO it is confusing, I leave the decision up to you. :)

contributors do not feel confident delivering support for it. This artifact or feature might be deprecated and 
removed from future James releases. Users are strongly encouraged to contribute to the artifact development.

## Consequences

> Quality levels needs to be mentioned explicitly in the documentation, per artifact, and per feature.

Reviewer 4: ```suggestion
Quality levels need to be mentioned explicitly in the documentation, per artifact, and per feature.
```
or
```suggestion
Quality level needs to be mentioned explicitly in the documentation, per artifact, and per feature.
```
